### DISCUSSION TITLE
Hardware Class

### DISCUSSION THREAD

**Discussion by ryanmrichard** [NWChemEx/ParallelZone#2] [Ideas]
The hardware classes are envisioned as an entire hierarchy. At the bottom we have the `HardwareComponent` class:
```.cpp
class Hardware {
public:
    // Prints a descriptive name of the hardware
    std::string name();
};
```

We at the moment distinguish two pieces of `Hardware`: things that can compute and things that store data. The former are modeled by
the `Processor` class and the latter by the `Storage` class:

```.cpp
struct Processor : HardwareComponent {
    // At the moment mainly for GPUs, but could in theory also be used for CPUs
    auto instruction_set();

    // How many threads does the device have? May need to move up to CPU/GPU
    // to deal with GPU hierarchy
    size_type nthreads();
};

struct Storage : HardwareComponent{
    /// How much space is free still?
    size_type free_space();

    /// How much total space is there?
    std::size_t total_sace();

    /// How fast is writing?
   size_type write_speed();

   /// How fast is reading?
   size_type read_speed();
};

```
The `Processor` class in theory could be home to a common thread-level tasking API, but for right now it mainly tells you the number of threads you have and can be used to determine what "instruction set" the processor uses (I'm thinking CUDA vs HIP vs SyCL, which aren't really instruction sets so other names are welcome). The `Storage` class is a bit more substantial and keeps track of how much memory is
in use and also allows you to select storage locations based on access speed. In theory we could also have the `Storage` class provide an allocator for the memory space.

From `Processor` and `Storage` we derive a series of classes which are more-or-less just strong types. 

```.cpp
struct CPU : Processor {};
struct GPU : Processor {};

struct Disk : Storage {
   std::iostream file(std::string name);
};
struct RandomAccessMemory : Storage {};
struct UnifiedMemory : Storage {};
```
At the moment the only real functionality added by these derived classes is the ability to open a file on a `Disk` instance.

**Comment by twindus** [comment]
Information from Anthony Danalis about PAPI routines:
PAPI_get_hardware_info()
papi_print_header()
print_header.c
